### PR TITLE
feat: session expiry

### PR DESCRIPTION
**Pull request by SanketD92**
Issue #, if available:
N/A

Description of changes:
This is the short-term fix for the ID token expiration bug reported by our customers. Currently the access tokens being generated (by Cognito or otherwise) have an expiration time of 60min, while the default inactivity-warning message has a threshold of 30min. This bug is reported to be independent of inactivity.

This PR adds a ForceLogout component keeping track of user's JWT token expiration and auto logout.

Checklist:

- [x] Have you successfully deployed to an AWS account with your changes?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully tested with your changes locally?

AS review ticket id:

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: addons/addon-base-ui/packages/base-ui/src/AppContainer.js [hunk 1]
```diff
@@ -58,8 +58,20 @@ class AppContainer extends Component {
       return false; // This will stop lodash from continuing the forEach loop
     });
 
+    plugins = _.reverse(pluginRegistry.getPluginsWithMethod('app-component', 'getForceLogoutComponent') || []);
+    let ForceLogout = () => <></>;
+    // We ask each plugin in reverse order if they have the ForceLogout component
+    _.forEach(plugins, plugin => {
```
Comment (nguyen102) [inline addons/addon-base-ui/packages/base-ui/src/AppContainer.js:64]: This code is duplicated three times. I think there's an opportunity to DRY it up by making that code a private method. 

Then we can just call `getComponent(component)`

File: addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  componentDidMount() {
+    this.timer = setInterval(() => {
+      const expirationInMins = this.tokenExpiresIn();
+      // eslint-disable-next-line react/no-unused-state
+      if (expirationInMins > 0) {
+        this.setState({ tokenActive: true });
+      } else {
+        this.setState({ tokenActive: false });
+      }
+    }, 1000);
```
Comment (nguyen102) [inline addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js:52]: I think this can be simplified by changing `tokenExpiresIn` to be `hasTokenExpired`. The method returns a boolean and we can use that to determine the state.
```
const tokenExpired = this.hasTokenExpired();
this.setState({ tokenExpired})
```

**Review by nguyen102 — COMMENTED** [review]
Overall looks good. Couple of small comments

File: addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  get modalOpen() {
+    return !this.state.tokenActive;
+  }
+
+  componentDidMount() {
+    this.timer = setInterval(() => {
+      const expirationInMins = this.tokenExpiresIn();
+      // eslint-disable-next-line react/no-unused-state
+      if (expirationInMins > 0) {
+        this.setState({ tokenActive: true });
```
Comment (hatimkhan) [inline addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js]: Any reason why you are mixing MobX state management with this.setState()?  
In general, there is no need to use this.setState() otherwise you need to worry about implementing shouldUpdate() correctly.  Using MobX makes this easier.  Instead of this.setState{{ tokenActive }).  Just declare 'tokenActive' as an observable instance variables (just like how you declared 'internvalId'.  Then use runInAction() when you need to update the tokenActive value.

File: addons/addon-base-ui/packages/base-ui/src/AppContainer.js (hunk 1 above)
Comment (nguyen102) [reply addons/addon-base-ui/packages/base-ui/src/AppContainer.js:64]: Looks like this approach involves a double for loop, which makes the code harder to read. Resolving this comment.

File: addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js (hunk 3 above)
Comment (nguyen102) [reply addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js]: We don't actually have to implement `shouldComponentUpdate`, right? It's usually only implemented if we want to optimize performance and don't want to rerender the REACT component on every `state` change. In our case the `state` should only change when the token has expire.

Comment (SanketD92) [reply addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js]: Good point. Made the changes.

File: addons/addon-base-ui/packages/base-ui/src/AppContainer.js (hunk 1 above)
Comment (SanketD92) [reply addons/addon-base-ui/packages/base-ui/src/AppContainer.js:64]: Per conversation, skipping this for readability.

File: addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js (hunk 3 above)
Comment (hatimkhan) [reply addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js]: @nguyen102  here are my thoughts:
With MobX,  it automatically provides an extremely efficient implementation of shouldComponentUpdate(), this way we never have to make a decision of when we should implement the shouldComponentUpdate().  It is better not to mix setState() if we are already using MobX to manage the state of this component.

File: addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    clearInterval(this.timer);
+    return this.authentication.logout({ autoLogout: true });
+  };
+
+  hasTokenExpired = () => {
+    let minutesRemaining;
+    try {
+      const idToken = localStorage.getItem('appIdToken');
+      const decodedIdToken = jwtDecode(idToken);
+      const expiresAt = _.get(decodedIdToken, 'exp', 0) * 1000;
```
Comment (nguyen102) [inline addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js:69]: this can just be `return _.get(decodedIdToken, 'exp', Date.now()) >= Date.now())`

File: addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    return this.props.authentication;
+  }
+
+  get modalOpen() {
+    return !this.tokenActive;
+  }
+
+  componentDidMount() {
+    this.timer = setInterval(() => {
+      const expirationInMins = this.hasTokenExpired();
```
Comment (nguyen102) [inline addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js]: The return value is a boolean now, so we can just do
```
runInAction(() => {
          this.tokenActive = this.hasTokenExpired();
        });
```

**Review by nguyen102 — CHANGES_REQUESTED** [review]

File: addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js (hunk 4 above)
Comment (SanketD92) [reply addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js:69]: Wanted to still keep this inside the try-catch block in case the token was not received at all, or if it was corrupted.

File: addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js (hunk 5 above)
Comment (SanketD92) [reply addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js]: Ah yes, had to revert my changes for lock file merge conflicts with develop. Pushed the fix.

File: addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js (hunk 4 above)
Comment (nguyen102) [reply addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js:69]: I was thinking something like this. Thoughts?
```
hasTokenExpired = () => {
    try {
      const idToken = localStorage.getItem('appIdToken');
      const decodedIdToken = jwtDecode(idToken);
      return _.get(decodedIdToken, 'exp', Date.now()) >= Date.now())
    } catch (e) {
      return false;
    }
  };
  ```

File: addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js (hunk 3 above)
Comment (nguyen102) [reply addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js]: @hatimkhan Yeah, I agree. If we're handling local state in a class, we should use either `mobx` or `REACT state`. I personally prefer to use REACT state if we're just handling local state for a component since REACT is already built to handle that.

In terms of worrying about `shouldComponentUpdate`, from what I've read and experienced, you rarely need to implement that method. Most of the time, the default REACT implementation for `shouldComponentUpdate` is sufficient. You only need to implement that method if your component's `render` method is very large and results in long rendering time if it were to be rerender often. 

So basically the heuristic I follow is not to implement `shouldComponentUpdate` unless I see a performance impact. Seeing a performance impact on the UI is rare.

More info about this can be read [here](http://jamesknelson.com/should-i-use-shouldcomponentupdate/)

Comment (hatimkhan) [reply addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js]: @nguyen102 I like your heuristic, thanks for sharing the link. Since I started using MobX, I never had to think about shouldUpdateComponent at all, with MobX we never need to write one, even for data intensive/expensive components, so you completely eliminate the need for heuristic.

File: addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js (hunk 4 above)
Comment (SanketD92) [reply addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js:69]: We can do it this way too. I originally wanted to convey to the reader computation is being done by the number of minutes left until token expiration.

**Comment by SanketD92** [comment]
Closing PR. CodeCov acting weird.
